fcrepo-search-impl/src/main/java/org/fcrepo/search/impl/DbSearchIndexImpl.java

```diff
@@ -346,33 +376,69 @@ private void deleteRdfTypeAssociations(final Long resourceId) {
                 deleteParams);
     }
 
-    private ArrayList<Long> findOrCreateRdfTypesInDb(final List<URI> rdfTypes) {
-        final var jdbcInsertRdfTypes = new SimpleJdbcInsert(this.jdbcTemplate.getJdbcTemplate());
-        jdbcInsertRdfTypes.withTableName(SEARCH_RDF_TYPE_TABLE).usingGeneratedKeyColumns(
-                ID_COLUMN);
-        final var rdfTypeIds = new ArrayList<Long>();
-        for (var rdfTypeUri : rdfTypes) {
-            final var typeParams = new MapSqlParameterSource();
-            typeParams.addValue(RDF_TYPE_URI_PARAM, rdfTypeUri.toString());
-            final var results = jdbcTemplate.queryForList(SELECT_RDF_TYPE_ID,
-                    typeParams);
-            if (CollectionUtils.isEmpty(results)) {
-                final Number key = jdbcInsertRdfTypes.executeAndReturnKey(typeParams);
-                rdfTypeIds.add(key.longValue());
-            } else {
-                rdfTypeIds.add((long) results.get(0).get(ID_COLUMN));
+    private List<Long> findOrCreateRdfTypesInDb(final List<URI> rdfTypes) {
+        final List<String> rdfTypes_str = rdfTypes.stream().map(URI::toString).collect(Collectors.toList());
+
+        final List<RdfType> results = jdbcTemplate.query(SELECT_RDF_TYPE_ID,
+                Map.of(RDF_TYPE_URI_PARAM, rdfTypes_str), RDF_TYPE_ROW_MAPPER);
+        // List of existing type ids.
+        final var rdfTypeIds = results.stream().map(RdfType::getTypeId).collect(Collectors.toList());
+        // List of existing type uris.
+        final var rdfTypeUris = results.stream().map(RdfType::getTypeUri).collect(Collectors.toList());
+        // Type uris that don't already have a record. Needs to be a set to avoid inserting the same URI and
+        final var missingUris = rdfTypes_str.stream().filter(t -> !rdfTypeUris.contains(t))
+                .collect(Collectors.toSet());
```

Comment:
Doesn't matter that much because I suspect we're only ever dealing with a handful of types, but this can be optimized a little:

```java
final var rdfTypeIds = new ArrayList<Long>(rdfTypes.size());
final var rdfTypeUris = new HashSet<String>();

for (final var type : results) {
    rdfTypeIds.add(type.getTypeId());
    rdfTypeUris.add(type.getTypeUri());
}

final var missingUris = rdfTypes.stream().map(URI::toString).filter(t -> !rdfTypeUris.contains(t)).collect(Collectors.toSet());
```